File: mobile-app/packages/services/test/pro_upgrade_service_parity_test.dart
import 'dart:convert';
import 'package:fake_async/fake_async.dart';
import 'package:smwa_services/services.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

class _Flag {
  bool value = false;
}

void main() {
  test('ledger increments again after window expiry', () {
    fakeAsync((async) {
      var calls = 0;
      final ledger = ApiQuotaLedger(1, const Duration(milliseconds: 50));
      final client = MockClient((req) async {
        calls++;
        expect(req.url.toString(),
            'http://localhost:12111/v1/checkout/sessions');
        return http.Response(jsonEncode({'status': 'open'}), 200);
      });
      final flag = _Flag();
      final svc = ProUpgradeService((v) async => flag.value = v,
          ledger: ledger, client: client);

      svc.checkoutMock().then(expectAsync1((ok) => expect(ok, isTrue)));
      async.elapse(const Duration());
      expect(flag.value, isTrue);
      expect(ledger.isSafe(), isFalse);

      async.elapse(const Duration(milliseconds: 60));
      expect(ledger.isSafe(), isTrue);

      svc.checkoutMock().then(expectAsync1((ok) => expect(ok, isTrue)));
      async.elapse(const Duration());
      expect(calls, 2);
      expect(ledger.isSafe(), isFalse);
    });
  });

  test('returns false when ledger disallows', () {
    fakeAsync((async) {
      var called = false;
      final ledger = ApiQuotaLedger(0);
      final client = MockClient((req) async {
        called = true;
        return http.Response('', 200);
      });
      final flag = _Flag();
      final svc = ProUpgradeService((v) async => flag.value = v,
          ledger: ledger, client: client);

      svc.checkoutMock().then(expectAsync1((ok) => expect(ok, isFalse)));
      async.elapse(const Duration());
      expect(called, isFalse);
      expect(flag.value, isFalse);
    });
  });

  test('returns false on network error', () {
    fakeAsync((async) {
      final ledger = ApiQuotaLedger(1);
      final client = MockClient((req) async => http.Response('', 500));
      final flag = _Flag();
      final svc = ProUpgradeService((v) async => flag.value = v,
          ledger: ledger, client: client);

      svc.checkoutMock().then(expectAsync1((ok) => expect(ok, isFalse)));
      async.elapse(const Duration());
      expect(flag.value, isFalse);
    });
  });
}

File: web-app/tests/ProUpgradeServiceParity.test.ts
import { describe, it, expect, vi, afterEach } from 'vitest';
import { ProUpgradeService } from '../src/services/ProUpgradeService';
import { ApiQuotaLedger } from '../src/utils/ApiQuotaLedger';

class TestService extends ProUpgradeService {
  constructor(private ledger: ApiQuotaLedger, private update: (v: boolean) => void) {
    super();
  }
  async checkoutMock(): Promise<boolean> {
    if (!this.ledger.isSafe()) return false;
    const ok = await super.checkoutMock();
    if (ok) {
      this.ledger.increment();
      await this.update(true);
    }
    return ok;
  }
}

afterEach(() => {
  vi.restoreAllMocks();
  vi.useRealTimers();
});

describe('ProUpgradeService parity with Dart', () => {
  it('ledger increments again after window expiry', async () => {
    vi.useFakeTimers();
    const ledger = new ApiQuotaLedger(1, 50);
    const fetchMock = vi.fn().mockResolvedValue({ ok: true });
    global.fetch = fetchMock as any;
    const flag = { value: false };
    const svc = new TestService(ledger, v => (flag.value = v));

    const ok1 = await svc.checkoutMock();
    expect(ok1).toBe(true);
    expect(fetchMock).toHaveBeenCalledTimes(1);
    expect(flag.value).toBe(true);
    expect(ledger.isSafe()).toBe(false);

    vi.advanceTimersByTime(60);
    expect(ledger.isSafe()).toBe(true);

    const ok2 = await svc.checkoutMock();
    expect(ok2).toBe(true);
    expect(fetchMock).toHaveBeenCalledTimes(2);
    expect(ledger.isSafe()).toBe(false);
  });

  it('returns false when ledger disallows', async () => {
    const ledger = new ApiQuotaLedger(0);
    const fetchMock = vi.fn();
    global.fetch = fetchMock as any;
    const svc = new TestService(ledger, () => {});

    const ok = await svc.checkoutMock();
    expect(ok).toBe(false);
    expect(fetchMock).not.toHaveBeenCalled();
  });

  it('returns false on network error', async () => {
    const ledger = new ApiQuotaLedger(1);
    global.fetch = vi.fn().mockResolvedValue({ ok: false }) as any;
    const svc = new TestService(ledger, () => {});

    const ok = await svc.checkoutMock();
    expect(ok).toBe(false);
  });
});
